### PR TITLE
Fine control of static SSR

### PR DESCRIPTION
**Pull request by guardrex**
Addresses #31403
Addresses #28161

* Now we're cook'in with ***gas*** 🔥.
* There's a section here named *Client-side services fail to resolve during prerendering* that you can ignore. That section is merely moving down in the article, and there are no changes to it.
* The two sections for review are:
  * ***Apply a render mode in code*** and its two subsections
  * ***Fine control of render modes*** and its two subsections
* We might want to change the section heading name, *Fine control of render modes*, but I'm not sure yet what would be better. I'll 🤔 about it.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/components/render-modes.md](https://github.com/dotnet/AspNetCore.Docs/blob/d005f79d64c26e4541fc704c300a054afb66928e/aspnetcore/blazor/components/render-modes.md) | [ASP.NET Core Blazor render modes](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?branch=pr-en-us-31497) |
| [aspnetcore/blazor/security/server/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/d005f79d64c26e4541fc704c300a054afb66928e/aspnetcore/blazor/security/server/index.md) | [[Visual Studio](#tab/visual-studio)](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/security/server/index?branch=pr-en-us-31497) |

### CODE REVIEW

File: aspnetcore/blazor/components/render-modes.md
Comment (MackinnonBuck) [inline aspnetcore/blazor/components/render-modes.md:181]: Hm, I actually think 'programmatically' might be a better work here, since one could argue that all render modes are applied "in code".

Or maybe 'dynamically' would be a better word, since it implies that the render mode is determined at runtime?

Comment (MackinnonBuck) [reply aspnetcore/blazor/components/render-modes.md:181]: Thinking about this more, I don't want the word 'dynamically' to give the illusion that this has anything to do with the auto render mode. Maybe 'programmatically' is better after all. What do you think?

File: aspnetcore/blazor/components/render-modes.md
Comment (MackinnonBuck) [inline aspnetcore/blazor/components/render-modes.md:188]: I'm not totally understanding why these are fundamentally different scenarios. Isn't the only difference between these what goes in the implementation for `RenderModeForPage`? i.e., in scenario 1, you decide the render mode by looking at the request path, and in scenario 2, you decide the render mode via reflection. If that's the case, could the code samples and descriptions for these be combined into one, where the potential ways to decide the render mode are explained in a single section? I could be missing something here, so let me know if I am!

**Review by MackinnonBuck — COMMENTED** [review]
Looks good so far!

I still need to do a detailed review, but I'm waiting on that because there's a chance this feedback results in consolidating some content, and I think it would be better to wait until that happens, if it does! 🙂

File: aspnetcore/blazor/components/render-modes.md
Comment (guardrex) [reply aspnetcore/blazor/components/render-modes.md:188]: It's often best in complex use cases to format the guidance around the goal that the dev is trying to achieve. It's often critical to adopt more of a tutorial-type of approach for the new-to-programming, new-to-.NET, and new-to-Blazor folks. 

In addition to the delta on how `RenderModeForPage` is set, there are additional considerations for the force-loading layout and what the individual components need to do in regard to use of the layout and setting the render mode.

When you see me write out scenarios in this degree of detail, it's usually because I found it confusing myself and detect that readers would benefit from more of step-by-step, detailed explanation and set of examples. It's long, yes, but it's utterly clear what's going on and what the dev needs to do. If we aren't clear, it will upset readers 😡 and they'll come for the 🦖 with sharpened *TEETH* and *CLAWS* 👹😨😆. I think we'll head off doc issues by explaining it out.

File: aspnetcore/blazor/components/render-modes.md
Comment (guardrex) [inline aspnetcore/blazor/components/render-modes.md:181]:
```suggestion
## Apply a render mode programatically
```

File: aspnetcore/blazor/components/render-modes.md
Comment (MackinnonBuck) [reply aspnetcore/blazor/components/render-modes.md:188]: Sounds great to me, thanks for clarifying! I definitely trust your judgement about what's clear to readers, so let's go with your original strategy of having it documented as two separate scenarios 🙂

Comment (guardrex) [reply aspnetcore/blazor/components/render-modes.md:188]: Cool ... you still need time to fully analyze this?

File: aspnetcore/blazor/components/render-modes.md
Comment (MackinnonBuck) [inline aspnetcore/blazor/components/render-modes.md:581]: > In the `App` component, any request for a component in the `Account` folder applies a `null` render mode, which enforces static SSR.

Just a small note about this: A `null` render mode is effectively the same as not specifying one at all, which results in the component inheriting its parent's render mode. In this case, the `App.razor` component gets rendered using static server rendering, so a `null` render mode makes the `<Routes>` component also use static server rendering. Whereas, if a `null` render mode were to be specified for a child component whose parent was interactive, the child would inherit the same interactive render mode.

All that to say, it's not always the case that `null` enforces static SSR, it just happens to behave that way in this example. Not sure whether it's worth calling out that caveat (maybe it would just create more confusion), so I'll leave it up to you to decide! Just want to make sure that readers don't see this statement and think that they can apply `null` as a render mode _anywhere_ to switch from interactive rendering to static server rendering.

File: aspnetcore/blazor/components/render-modes.md
Comment (MackinnonBuck) [inline aspnetcore/blazor/components/render-modes.md:652]: Maybe there's a reason you didn't do this way, but would it make sense to just have a single ` ```razor ` block with a `@code` section rather than two different code blocks? Maybe that would make the example more copy-pastable.

**Review by MackinnonBuck — APPROVED** [review]
Approving now because I think this looks great - feel free to address my feedback however you like!

**Comment by guardrex** [comment]
Thanks @MackinnonBuck ... I'll make further updates based on your feedback on Friday.

File: aspnetcore/blazor/components/render-modes.md
Comment (guardrex) [reply aspnetcore/blazor/components/render-modes.md:581]: Yes ... great idea to add this. I placed it in an IMPORTANT note immediately after the paragraph with the remark. This is a *teachable moment*, to borrow an educational slogan, on this because we have the reader's focus on how this code is working and we're right in the thick of the whole render mode explanation in the article. Great idea to call it out.

**Pull request merged**
